### DISCUSSION TITLE
Dynamic List Sizes for Nested Objects

### DISCUSSION THREAD

**Discussion by StathisKons** [instancio/instancio#1175] [Q&A]
Hi, I'm trying to generate a list of `Foo` objects, where each `Foo` has multiple nested objects and lists. I want to customize the size of each list with custom logic so that each generated object has a different number of nested objects.

For example, given the classes:

```java
public class Foo {
  private List<Bar> bars;
  private List<Baz> bazs;
  // other fields...
}

public class Bar {
  private List<Oof> oofs;
  // other fields...
}
```
I tried the following approach:

```java
Instancio.ofList(Foo.class)
    .size(fooAmount)
    .generate(field(Foo.class, "bars"), gen -> gen.collection().size(getBarsAmount()))
    .generate(field(Bar.class, "oofs"), gen -> gen.collection().size(getOofsAmount()))
    .create();
```
However, the problem I'm encountering is that `getBarsAmount()` is only called once. As a result, every `Foo` object has the same number of `bars`, but I need each `Foo` object to call `getBarsAmount()` separately, so each instance has a different number of `bars`.

Is there a way to ensure that the size method is called for each individual object and not once for all of them? Any suggestions or advice would be greatly appreciated!

Feel free to ask for any clarifications if needed. Thanks!

**Comment by StathisKons** [comment]
My workaround was something like:
```java
Instancio.ofList(Foo.class)
	.size(fooAmount)
	.supply(field(Foo.class, "bars"), this::createBars)
	.create();

private createBars(){
	Instancio.ofList(Bar.class)
		.size(getBarsAmount)
		.supply(field(Bar.class, "oofs"), this::createOofs)
		.create();
}
```
but I feel like there must be a better solution, this also creates other problems, since `Foo` `Bar` and other classes have fields that must be unique, so generating those values with this workaround also becomes more complex

**Reply by armandino** [reply]
I see what you're trying to do. Unfortunately there's no concise way to do this. One option is to create a custom generator for `List<Foo>`. Here's an example:
```java
@ExtendWith(InstancioExtension.class)
class SampleTest {

    @Test
    void example() {
        final int initialListSize = 1;

        final List<Foo> results = Instancio.ofList(Foo.class)
                .size(4)
                .supply(field(Foo::bars), new FooListGenerator(initialListSize))
                .create();

        assertThat(results.get(0).bars()).hasSize(1);
        assertThat(results.get(1).bars()).hasSize(2);
        assertThat(results.get(2).bars()).hasSize(3);
        assertThat(results.get(3).bars()).hasSize(4);
    }


    record Foo(List<Bar> bars) {}
    record Bar(List<Oof> oofs) {}
    record Oof(String value) {}

    class FooListGenerator implements Generator<List<Foo>> {
        private int numberOfElements;

        FooListGenerator(int numberOfElements) {
            this.numberOfElements = numberOfElements;
        }

        @Override
        public List<Foo> generate(Random random) {
            // E.g. each subsequent list is larger than the previous by 1 element
            numberOfElements++;

            // returning empty list, instancio will it with objects
            return new ArrayList<>();
        }

        @Override
        public Hints hints() {
            return Hints.builder()
                    .with(CollectionHint.builder()
                            .generateElements(numberOfElements) // number of list elements to generate
                            .build())
                    .build();
        }
    }
}
```

You can check out the docs for more details:

- https://www.instancio.org/user-guide/#custom-generators
- https://javadoc.io/doc/org.instancio/instancio-core/latest/org/instancio/generator/hints/CollectionHint.html

It is somewhat verbose though. I'm not sure if the added complexity is worth it. It depends on your test case. If not, it might be easier to construct the object manually.
